src/backend/optimizer/plan/planner.c

```diff
@@ -1866,6 +1906,352 @@ grouping_planner(PlannerInfo *root, double tuple_fraction)
+						}
+
+						context->mv_pure_vars = mv_pure_vars;
+						context->mv_pure_vars_index = mv_pure_vars_index;
+						context->mv_nonpure_vars_index = mv_nonpure_vars_index;
+						context->mv_tlist = mvRelQueryTree->targetList;
+						context->mv_query_tlist = mvQuery->targetList;
+						context->has_unmatched = false;
+
+						/* Sort nonpure vars expression, prepare for Greedy Algorithm. */
+						aqumv_process_nonpure_vars_expr(context);
+
+						List	*mv_final_tlist = NIL; /* Final target list we want to rewrite to. */
+
+						/*
+						 * Process and rewrite target list, return false if failed.
+						 */
+						if(!aqumv_process_targetlist(context, parse->targetList, &mv_final_tlist))
+							continue;
+
+						/*
+						 * AQUMV
+						 * Process all quals to conjunctive normal form.
+						 * 
+						 * We assume that the selection predicates of view and query expressions
+						 * have been converted into conjunctive normal form(CNF) before we process
+						 * them.
+						 * AQUMV_MVP: no having quals now.
+						 */
+						preprocess_qual_conditions(subroot, (Node *) mvQuery->jointree);
+
+						List *post_quals = NIL;
+						/*
+						 * Process quals, return false if failed. 
+						 * Else, post_quals are filled if there were. 
+						 * Like process target list, post_quals is used later to see if we could
+						 * rewrite and apply it to mv relation.
+						 */
+						if(!aqumv_process_from_quals(parse->jointree->quals, mvQuery->jointree->quals, &post_quals))
+							continue;
+
+						/* Rewrite post_quals, return false if failed. */
+						post_quals = (List *)aqumv_adjust_sub_matched_expr_mutator((Node *)post_quals, context);
+						if (context->has_unmatched)
+							continue;
+
+						/*
+						 * Here! We succeed to rewrite a new SQL.
+						 * Begin to replace all guts.
+						 */
+						mvQuery->targetList = mv_final_tlist;
+
+						/*
+						 * AQUMV
+						 * NB: Update processed_tlist again in case that tlist has been changed. 
+						 */
+						preprocess_targetlist(subroot);
+
+						/* Rewrite with mv's query tree*/
+						mvrte->relkind = RELKIND_MATVIEW;
+						mvrte->relid = matviewRel->rd_rel->oid;
+						/*
+						 * AQUMV_FIXME_MVP
+						 * Not sure where it's true from actions even it's not inherit tables.
+						 */
+						mvrte->inh = false;
+						mvQuery->rtable = list_make1(mvrte); /* rewrite to SELECT FROM mv itself. */
+						mvQuery->jointree->quals = (Node *)post_quals; /* Could be NULL, but doesn'y matter for now. */
+
+						/*
+						 * AQUMV
+						 * Build a plan of new SQL.
+						 * AQUMV is cost-based, let planner decide which is better.
+						 * AQUMV_FIXME_MVP:
+						 * 	no qp_callback function now.
+						 * 	replcace one-by-one?
+						 */
+						mv_final_rel = query_planner(subroot, NULL, NULL);
+
+						/* AQUMV_FIXME_MVP: TODO limit caluse and startup_cost. */
+						if (mv_final_rel->cheapest_total_path->total_cost < current_rel->cheapest_total_path->total_cost)
```

Comment:
use `compare_path_costs_fuzzily`?

Reply:
Seems reasonable.  I'm considering if we use this function or only with fuzzy  part because that function also take startup_cost into account when comparing, but we haven't consider limit clause in MVP0.

Reply:
BTW... a typo `caluse` -> `caluse`